doc/source/dev_guide/satpy_internals.rst

```diff
@@ -0,0 +1,152 @@
+      resolution:
+        transitive: true
+      calibration:
+        enum:
+            - reflectance
+            - brightness_temperature
+            - radiance
+            - counts
+      modifiers:
+        required: true
+        default: []
+        type: !!python/name:satpy.dataset.ModifierTuple
+
+To create a new set, the user can provide indications in the relevant yaml file.
+It has to be provided in header of the reader configuration file, under the `reader`
+section, as `data_identification_keys`. Each key under this is the name of relevant
+metadata key that will used to find relevant information in the attributes of the data
+arrays. Under each of this, a few options are available:
+
+ - `required`: if the item is required, False by default
+ - `type`: the type to use. More on this further down.
+ - `enum`: if the item has to be limited to a finite number of options, an enum can be used.
+   Be sure to place the options in the order of preference, with the most desirable option on top.
+ - `default`: the default value to assign to the item if nothing (or None) is provided. If this
+   option isn't provided, the key will simply be omited if it is not present in the attrs or if it
+   is None. It will be passed to the type's `convert` method if available.
+ - `transitive`: whether the key is to be passed when looking for dependencies. Here for example,
+   a composite that has to be at a certain resolution will pass this resolution requirement to its
+   dependencies.
+
+
+If the definition of the metadata keys need to be done in python rather than in a yaml file, it will
+be a dictionary very similar to the yaml code. Here is the same example as above in python:
+
+  .. code-block:: python
+
+    from satpy.dataset import WavelengthRange, ModifierTuple
+
+    id_keys_config = {'name': {
+                          'required': True,
+                      },
+                      'wavelength': {
+                          'type': WavelengthRange,
+                      },
+                      'resolution': None,
+                      'calibration': {
+                          'enum': [
+                              'reflectance',
+                              'brightness_temperature',
+                              'radiance',
+                              'counts'
+                              ]
+                      },
+                      'modifiers': {
+                          'required': True,
+                          'default': ModifierTuple(),
+                          'type': ModifierTuple,
+                      },
+                      }
+
+Types
+~~~~~
+Types are classes that implement a type to be used as value for metadata in the `DataID`. They have
+to implement a few methods:
+
+ - a `convert` class method that returns it's argument as an instance of the class
+ - `__hash__`, `__eq__` and `__ne__` methods
+ - a `distance` method the tells how "far" an instance of this class is from it's argument.
```

Comment:
Is there any example of how to implement such a class and in what context it may be useful?

Reply:
Fixed through documentation